File: win-prog/ProgWinVB/Tree View and List View/DirectoriesAndFiles/DirectoryTreeView.vb
'--------------------------------------------------
' DirectoryTreeView.vb (c) 2002 by Charles Petzold
'--------------------------------------------------
Imports System
Imports System.Drawing
Imports System.IO
Imports System.Windows.Forms

Class DirectoryTreeView
    Inherits TreeView

    Sub New()
        ' Make a little more room for long directory names.
        Width *= 2

        ' Get images for tree.
        ImageList = New ImageList()
        ImageList.Images.Add(New Bitmap(Me.GetType(), "35FLOPPY.BMP"))
        ImageList.Images.Add(New Bitmap(Me.GetType(), "CLSDFOLD.BMP"))
        ImageList.Images.Add(New Bitmap(Me.GetType(), "OPENFOLD.BMP"))

        ' Construct tree.
        RefreshTree()
    End Sub

    Sub RefreshTree()
        ' Turn off visual updating and clear tree.
        BeginUpdate()
        Nodes.Clear()

        ' Make disk drives the root nodes. 
        Dim astrDrives() As String = Directory.GetLogicalDrives()
        Dim str As String
        For Each str In astrDrives
            Dim tnDrive As New TreeNode(str, 0, 0)
            Nodes.Add(tnDrive)
            AddDirectories(tnDrive)
            If str = "C:\" Then
                SelectedNode = tnDrive
            End If
        Next str
        EndUpdate()
    End Sub

    Private Sub AddDirectories(ByVal tn As TreeNode)
        tn.Nodes.Clear()
        Dim strPath As String = tn.FullPath
        Dim dirinfo As New DirectoryInfo(strPath)
        Dim adirinfo() As DirectoryInfo

        ' Avoid message box reporting drive A has no diskette!
        If Not dirinfo.Exists Then Return

        Try
            adirinfo = dirinfo.GetDirectories()
        Catch
            Return
        End Try

        Dim di As DirectoryInfo
        For Each di In adirinfo
            Dim tnDir As New TreeNode(di.Name, 1, 2)
            tn.Nodes.Add(tnDir)

            ' We could now fill up the whole tree with this statement:
            '         AddDirectories(tnDir)
            ' But it would be too slow. Try it!

        Next di
    End Sub

    Protected Overrides Sub OnBeforeExpand(ByVal tvcea As TreeViewCancelEventArgs)
        MyBase.OnBeforeExpand(tvcea)
        BeginUpdate()

        Dim tn As TreeNode
        For Each tn In tvcea.Node.Nodes
            AddDirectories(tn)
        Next tn
        EndUpdate()
    End Sub
End Class

File: win-prog/ProgWinCS/Tree View and List View/DirectoriesAndFiles/DirectoryTreeView.cs
//---------------------------------------------
// DirectoryTreeView.cs © 2001 by Charles Petzold
//---------------------------------------------
using System;
using System.Drawing;
using System.IO;
using System.Windows.Forms;

class DirectoryTreeView: TreeView
{
     public DirectoryTreeView()
     {
               // Make a little more room for long directory names.

          Width *= 2;

               // Get images for tree.

          ImageList = new ImageList();
          ImageList.Images.Add(new Bitmap(GetType(), "35FLOPPY.BMP"));
          ImageList.Images.Add(new Bitmap(GetType(), "CLSDFOLD.BMP"));
          ImageList.Images.Add(new Bitmap(GetType(), "OPENFOLD.BMP"));

               // Construct tree.

          RefreshTree();
     }
     public void RefreshTree()
     {
               // Turn off visual updating and clear tree.

          BeginUpdate();
          Nodes.Clear();

               // Make disk drives the root nodes. 

          string[] astrDrives = Directory.GetLogicalDrives();

          foreach (string str in astrDrives)
          {
               TreeNode tnDrive = new TreeNode(str, 0, 0);
               Nodes.Add(tnDrive);
               AddDirectories(tnDrive);

               if (str == "C:\\")
                    SelectedNode = tnDrive;
          }
          EndUpdate();
     }
     void AddDirectories(TreeNode tn)
     {
          tn.Nodes.Clear();

          string          strPath = tn.FullPath;
          DirectoryInfo   dirinfo = new DirectoryInfo(strPath);
          DirectoryInfo[] adirinfo;

          try
          {
               adirinfo = dirinfo.GetDirectories();
          }
          catch
          {
               return;
          }

          foreach (DirectoryInfo di in adirinfo)
          {
               TreeNode tnDir = new TreeNode(di.Name, 1, 2);
               tn.Nodes.Add(tnDir);

               // We could now fill up the whole tree with this statement:
               //        AddDirectories(tnDir);
               // But it would be too slow. Try it!
          }
     }
     protected override void OnBeforeExpand(TreeViewCancelEventArgs tvcea)
     {
          base.OnBeforeExpand(tvcea);

          BeginUpdate();

          foreach (TreeNode tn in tvcea.Node.Nodes)
               AddDirectories(tn);

          EndUpdate();
     }
}